Question2: Debt Prediction

    
    In this question, we work with another dataset from the textbook of "An Introduction to Statistical Learning."



a- Read the dataset file “Credit.csv” (you should download it from CSNS), and assign it to a
Pandas DataFrame.

In [1]:
import numpy as np
import pandas as pd

credit_df = pd.read_csv('C:/RiddhiDocument/MS Computer Science/CS-4661 Data Sceince/Homework/HW3/Credit.csv')


print(credit_df)

      Income  Limit  Rating  Cards  Age  Education  Married  Balance
0     14.891   3606     283      2   34         11        1      333
1    106.025   6645     483      3   82         15        1      903
2    104.593   7075     514      4   71         11        0      580
3    148.924   9504     681      3   36         11        0      964
4     55.882   4897     357      2   68         16        1      331
5     80.180   8047     569      4   77         10        0     1151
6     20.996   3388     259      2   37         12        0      203
7     71.408   7114     512      2   87          9        0      872
8     15.125   3300     266      5   66         13        0      279
9     71.061   6819     491      3   41         19        1     1350
10    63.095   8117     589      4   30         14        1     1407
11    15.045   1311     138      3   64         16        0        0
12    80.616   5308     394      1   57          7        1      204
13    43.682   6922     511      1

b- Check out the dataset. The “Credit” dataset includes “balance” column (average credit
card debt for a number of individuals) as target, as well as several features: age, cards
(number of credit cards), education (years of education), income (in thousands of dollars),
limit (credit limit), marital status, and rating (credit rating).

In [2]:
credit_df[0::10]

Income  Limit  Rating  Cards  Age  Education  Married  Balance
0     14.891   3606     283      2   34         11        1      333
10    63.095   8117     589      4   30         14        1     1407
20    17.700   2860     235      4   63         16        0       89
30    34.142   5666     413      4   47          5        1      863
40    34.950   3327     253      3   54         14        0       50
50    36.362   5183     376      3   49         15        1      654
60    35.510   5198     364      2   35         20        0      631
70    24.889   3954     318      4   75         12        1      357
80    27.369   3449     288      3   40          9        1      162
90    20.191   5767     431      4   42         16        1     1023
100   21.153   3736     256      1   41         11        0      298
110   34.537   3271     250      3   57         17        1       47
120   27.241   1402     128      2   67         15        1        0
130   23.793   3821     281      4   56         12        1      868
140   39.831   6045     459      3   32         12        1     1425
150   63.931   5728     435      3   28         14        1      581
160   12.238   4865     381      5   67         11        0      836
170   34.772   2021     167      3   57          9        0        0
180   10.635   3584     294      5   69         16        1      423
190   55.187   5352     385      4   50         17        1      538
200   23.949   5343     383      2   40         18        1      829
210   24.543   3206     243      2   62         12        1       95
220   44.847   5765     437      3   53         13        0     1246
230   33.214   5137     387      3   59          9        0      661
240   29.705   3351     262      5   71         14        1      148
250   10.363   2430     191      2   47         18        1        0
260   67.937   5184     383      4   63         12        1      345
270   15.866   3085     217      1   39         13        0      136
280   53.401   5319     377      3   35         12        0      541
290   26.370   3235     268      5   78         11        1      159
300   21.786   4632     355      1   50         17        1      580
310   33.694   4891     369      1   52         16        0     1036
320   16.279   1160     126      3   78         13        1        5
330   20.405   4543     329      2   72         17        0     1054
340   27.794   3807     301      4   35          8        1      320
350   30.002   1561     155      4   70         13        1        0
360   53.566   5891     434      4   82         10        0      712
370   35.610   6135     466      4   40         12        0      992
380  115.123   7760     538      3   83         14        0      661
390  135.118  10578     747      3   81         15        1     1393

c- Generate the feature matrix and target vector (target is “balance” in this dataset). Then,
normalize (scale) the features (note: don’t normalize the target vector!). To normalize
the data, you can simply use preprocessing.scale(X) from sklearn.

In [3]:
featu_col = ['Income','Limit','Rating','Cards','Age','Education','Married']

X = credit_df[featu_col]
y = credit_df['Balance']


print(X.head())
print(y.head())

#print(X) # Print whole matrix
#print(y)

    Income  Limit  Rating  Cards  Age  Education  Married
0   14.891   3606     283      2   34         11        1
1  106.025   6645     483      3   82         15        1
2  104.593   7075     514      4   71         11        0
3  148.924   9504     681      3   36         11        0
4   55.882   4897     357      2   68         16        1
0    333
1    903
2    580
3    964
4    331
Name: Balance, dtype: int64


In [4]:
def featu_scalling(X):
    return X/max(X)


credit_df['Income'] = featu_scalling(credit_df['Income'])
credit_df['Limit'] = featu_scalling(credit_df['Limit'])
credit_df['Rating'] = featu_scalling(credit_df['Rating'])
credit_df['Cards'] = featu_scalling(credit_df['Cards'])
credit_df['Age'] = featu_scalling(credit_df['Age'])
credit_df['Education'] = featu_scalling(credit_df['Education'])

credit_df

Income     Limit    Rating     Cards       Age  Education  Married  \
0    0.079787  0.259182  0.288187  0.222222  0.346939       0.55        1   
1    0.568090  0.477611  0.491853  0.333333  0.836735       0.75        1   
2    0.560418  0.508517  0.523422  0.444444  0.724490       0.55        0   
3    0.797947  0.683102  0.693483  0.333333  0.367347       0.55        0   
4    0.299420  0.351973  0.363544  0.222222  0.693878       0.80        1   
5    0.429611  0.578380  0.579430  0.444444  0.785714       0.50        0   
6    0.112498  0.243513  0.263747  0.222222  0.377551       0.60        0   
7    0.382610  0.511320  0.521385  0.222222  0.887755       0.45        0   
8    0.081041  0.237188  0.270876  0.555556  0.673469       0.65        0   
9    0.380751  0.490117  0.500000  0.333333  0.418367       0.95        1   
10   0.338068  0.583411  0.599796  0.444444  0.306122       0.70        1   
11   0.080612  0.094228  0.140530  0.333333  0.653061       0.80        0   
12   0.431947  0.381514  0.401222  0.111111  0.581633       0.35        1   
13   0.234052  0.497520  0.520367  0.111111  0.500000       0.45        1   
14   0.102575  0.236541  0.273931  0.222222  0.765306       0.65        0   
15   0.107638  0.181485  0.203666  0.333333  0.581633       0.75        1   
16   0.287182  0.266945  0.291242  0.333333  0.744898       0.85        1   
17   0.195549  0.314670  0.345214  0.333333  0.704082       0.75        1   
18   0.265600  0.458851  0.456212  0.111111  0.285714       0.45        1   
19   0.225463  0.476245  0.487780  0.222222  0.448980       0.45        0   
20   0.094838  0.205563  0.239308  0.444444  0.642857       0.80        0   
21   0.200114  0.458420  0.466395  0.111111  0.734694       0.85        0   
22   0.107713  0.189104  0.216904  0.333333  0.622449       0.50        1   
23   0.343062  0.372242  0.405295  0.555556  0.489796       0.40        1   
24   0.057557  0.126285  0.158859  0.333333  0.581633       0.75        0   
25   0.075495  0.310717  0.331976  0.555556  0.255102       0.80        1   
26   0.227563  0.260548  0.294297  0.666667  0.448980       0.60        0   
27   0.175708  0.325882  0.339104  0.222222  0.448980       0.80        0   
28   1.000000  0.964134  0.966395  0.222222  0.418367       0.70        1   
29   0.143666  0.403292  0.418534  0.444444  0.561224       0.80        0   
..        ...       ...       ...       ...       ...        ...      ...   
370  0.190801  0.440955  0.474542  0.444444  0.408163       0.60        0   
371  0.209587  0.154532  0.176171  0.444444  0.765306       0.75        0   
372  0.105994  0.271832  0.298371  0.222222  0.469388       0.80        0   
373  0.296902  0.384820  0.390020  0.222222  0.377551       0.80        1   
374  0.157528  0.347876  0.374745  0.333333  0.775510       0.90        1   
375  0.112380  0.407748  0.420570  0.555556  0.448980       0.80        1   
376  0.469502  0.515130  0.524440  0.222222  0.469388       0.50        0   
377  0.150798  0.112628  0.144603  0.333333  0.520408       0.50        1   
378  0.103674  0.355135  0.372709  0.111111  0.336735       0.95        1   
379  0.285629  0.205563  0.217923  0.111111  0.857143       0.50        1   
380  0.616838  0.557752  0.547862  0.333333  0.846939       0.70        0   
381  0.545388  0.577086  0.584521  0.222222  0.857143       0.55        1   
382  0.133009  0.394954  0.416497  0.111111  0.336735       0.45        0   
383  0.076578  0.235319  0.287169  1.000000  0.653061       0.45        1   
384  0.107633  0.134407  0.183299  0.333333  0.775510       0.80        0   
385  0.141453  0.405376  0.405295  0.333333  0.591837       0.75        0   
386  0.103159  0.264716  0.292261  0.444444  0.581633       0.50        0   
387  0.088564  0.097535  0.128310  0.333333  0.632653       0.45        0   
388  0.202953  0.490692  0.490835  0.222222  0.816327       0.65        0   
389  0.449800  0.510314  0.512220  0.222222  0.448980       0.90        0   
390

d- Split the dataset into testing and training sets with the following parameters:
test_size=0.24, random_state=4.

In [5]:
from sklearn.model_selection import train_test_split

X_train,X_test,y_train,y_test = train_test_split(X,y,test_size=0.24,random_state=4)

print("X_train.shape :",X_train.shape)
print("y_train.shape :",y_train.shape)


print("X_test.shape :",X_test.shape)
print("y_test.shape :",y_test.shape)


X_train.shape : (304, 7)
y_train.shape : (304,)
X_test.shape : (96, 7)
y_test.shape : (96,)


e- Use Linear Regression to train a linear model on the training set. Check the coefficients of
the linear regression model. Which feature is the most important? Which feature is the
least important?

In [6]:
from sklearn.linear_model import LinearRegression

myLinreg = LinearRegression()
myLinreg.fit(X_train,y_train)



LinearRegression(copy_X=True, fit_intercept=True, n_jobs=1, normalize=False)

In [7]:
print("\n Linear Regression Intercept : ",myLinreg.intercept_)
print("\n Linear Refression Coefficient : ",myLinreg.coef_)


 Linear Regression Intercept :  -429.5609688670513

 Linear Refression Coefficient :  [ -7.52790609   0.05725541   3.09667887  10.7488698   -1.72874937
   0.53768504 -41.47581319]


Which feature is the most important? Which feature is the
least important?

ANS : Linear Regression have most important feature is 10.7488698 it is Cards , 
      and Linear Regression have least important feature is -1404.96322442 it is Age.

f- Predict “balance” for the users in testing set. Then, compare the predicted balance with
the actual balance by calculating and reporting the RMSE (as we saw in lab tutorial 4).

In [8]:
y_prediction = myLinreg.predict(X_test)
print(y_prediction)

[ 1.40192822e+03  3.72390498e+02  3.09606997e+02  7.75587837e+02
  3.63641910e+02  9.54490329e+02  3.20745279e+02 -2.32591690e+02
  3.99394342e+02  6.01919205e+02 -2.52478552e+02  3.82362570e+02
  4.72524424e+02  7.15924136e+02  1.11080816e+01  6.76777695e+02
 -1.10205182e+02  2.34123414e+02 -2.17604376e+02  1.36106372e+03
  3.62716990e+02 -7.18441620e+01  3.46590403e+02 -3.48468421e+00
 -1.98897875e+02 -3.79979558e+01  7.87248445e+02  3.41641943e+02
  6.76645933e+02 -5.36248977e+01  7.73056436e+02  1.50270904e+01
  9.34623394e+02  3.80185173e+02  7.48107862e+02 -2.85720687e+02
  1.00583355e+03  5.07594510e+02 -1.40361418e+02  1.10973338e+03
  1.08474720e+03  3.98005654e+02  6.40559128e+02  9.08998220e+02
  6.99056037e+02  7.92919368e+02  3.82003132e+02  9.70862036e+02
  3.34390882e+02  1.16908059e+03  2.43605505e+02  2.78833488e+02
 -6.23362942e-01  3.85459830e+01  2.49931072e+02  8.25222693e+02
 -1.98555175e+02  5.71140218e+01  1.17285043e+03  9.81760238e+02
 -6.84711889e+00  2.70524

In [13]:
from sklearn  import metrics

mse = metrics.mean_squared_error(y_test,y_prediction)
rmse = np.sqrt(mse)
print("Calculating RMSE :",rmse)

Calculating RMSE : 161.51385491175319


g- Now, use 10-fold Cross-Validation to evaluate the performance of a linear regression in
predicting the balance. Thus, rather than splitting the dataset into testing and training,
use Cross-Validation to evaluate the regression performance. What is the RMSE when you
use cross validation?

In [14]:
from sklearn.cross_validation import cross_val_score

myLinreg = LinearRegression()
rmseList = cross_val_score(myLinreg,X,y,cv=10,scoring='neg_mean_squared_error')

print("RMSE List : ",rmseList)


RMSE List :  [-23646.90415342 -32003.04401232 -35462.64435619 -37327.60719635
 -14341.32205938 -33628.37104224 -31631.99317834 -12491.00334951
 -20749.61212176 -23204.9474346 ]


In [15]:
rmseListPosi = -rmseList
rmseList = np.sqrt(rmseListPosi)

print("RMSE positive List : ",rmseList )


RMSE positive List :  [153.7754992  178.89394627 188.31527914 193.20353826 119.75525901
 183.38039983 177.85385343 111.76315739 144.04725656 152.331702  ]


In [16]:
print("RMSE positive List : ",rmseList.mean())

RMSE positive List :  160.33198910744073
